### PR TITLE
Punish peers for duplicate GRANDPA neighbor messages

### PR DESCRIPTION
**Pull request by dmitry-markin**
Resolves https://github.com/paritytech/substrate/issues/12191.

The code already contains logic that doesn't allow sending GRANDPA neighbor messages with the same or decreasing set ID / round within a set / finalized number. This PR only enforces this upon reception and reports misbehaving peers.

Follow-up issue for a generalized case: https://github.com/paritytech/substrate/issues/12463.

### CODE REVIEW

File: client/finality-grandpa/src/communication/gossip.rs
Comment (bkchr) [inline client/finality-grandpa/src/communication/gossip.rs:546]: Maybe we could split this into multiple lines. 

```
let set_id_increased = update.set_id > peer.view.set_id;
let round_increased = update.set_id == peer.view.set_id && update.round > peer.view.round;
...
...
```

You see what I want.

**Review by altonen — APPROVED** [review]

File: client/finality-grandpa/src/communication/gossip.rs
Comment (dmitry-markin) [reply client/finality-grandpa/src/communication/gossip.rs]: Is it better now? I've also added check that `commit_finalized_height` cannot move backwards event if set ID or round has increased.

**Comment by bkchr** [comment]
@dmitry-markin please merge master to make CI happy ;)

And also please wait for @andresilva approval.

File: client/finality-grandpa/src/communication/gossip.rs
Comment (dmitry-markin) [reply client/finality-grandpa/src/communication/gossip.rs:548]: If the set_id or round has progressed, `commit_finalized_height` can stay the same, but can't go backwards. So here is the first check: `!finalized_height_decreased`. But if the set_id and round are the same, `commit_finalized_height` can only increase, so there is the second check `finalized_height_in_round_increased`.

Do I misunderstand something?

Comment (dmitry-markin) [reply client/finality-grandpa/src/communication/gossip.rs:548]: The last check `finalized_height_in_round_increased` is appended with `||`, so it can work on its own.

Comment (bkchr) [reply client/finality-grandpa/src/communication/gossip.rs:548]: Ahh yeah :facepalm: Sorry!

File: client/finality-grandpa/src/communication/gossip.rs
Comment (bkchr) [reply client/finality-grandpa/src/communication/gossip.rs:550]: Still I would like to see this ;)

Comment (dmitry-markin) [reply client/finality-grandpa/src/communication/gossip.rs:550]: Would it be better like `finalized_height_not_decreased`? `finalized_height_increased` is not quite the same as `!finalized_height_decreased` :)

Comment (bkchr) [reply client/finality-grandpa/src/communication/gossip.rs:550]: Yeah good point. `finalized_height_not_decreased` sounds fine!

Comment (dmitry-markin) [reply client/finality-grandpa/src/communication/gossip.rs]: Done!

File: client/finality-grandpa/src/communication/gossip.rs
Comment (andresilva) [inline client/finality-grandpa/src/communication/gossip.rs:557]: I think it might make sense to create a new misbehavior variant for this, maybe `DuplicateNeighborMessage`, which would have a lower punishment cost (maybe -100?). This is because we have logic to periodically rebroadcast neighbor messages in case there is no progress happening (https://github.com/paritytech/substrate/blob/master/client/finality-grandpa/src/communication/periodic.rs#L57-L60), so there might be situations where nodes legitimately send out a duplicate neighbor message and I think we want to reduce the reputation less harsh then when an actual invalid view change happens. This would still address the problem of someone spamming neighbor messages.

**Review by andresilva — CHANGES_REQUESTED** [review]

File: client/finality-grandpa/src/communication/gossip.rs
Comment (dmitry-markin) [reply client/finality-grandpa/src/communication/gossip.rs:557]: Should we align the period of rebroadcasting neighbor messages with punishment logic then? Like, punish only peers that spam us more frequently than the rebroadcast period?

Comment (andresilva) [reply client/finality-grandpa/src/communication/gossip.rs:557]: That would certainly be better than what I suggested! Keeping track of when we got the last neighbor message per peer and then if we receive a duplicate after `PERIODIC_REBROADCAST` it is valid, otherwise it's spam.

Comment (dmitry-markin) [reply client/finality-grandpa/src/communication/gossip.rs:560]: Done.

File: client/finality-grandpa/src/communication/gossip.rs
Comment (dmitry-markin) [inline client/finality-grandpa/src/communication/gossip.rs:1820]: I don't like this test:

1. It modifies internal structure that can change, and the test must be updated in this case.
2. It rolls clock backwards: on systems with unsigned time representation and 0 time corresponding to the system boot, if the test is run less than two minutes after boot, it'll fail.

The alternative I see is modifying `Peers::update_peer_state()` to accept time instant as an argument, but this will require measuring time on the calling side. This seems to be even worse.

Do you see any better options?

**Review by andresilva — APPROVED** [review]

**Review by altonen — APPROVED** [review]

**Comment by dmitry-markin** [comment]
bot merge

**Pull request merged**
